sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -683,7 +714,20 @@ abstract class HadoopFsRelation private[sql](
       }
     }
 
-    buildInternalScan(requiredColumns, filters, inputStatuses, broadcastedConf)
+    groupBucketFiles(inputStatuses).map { groupedBucketFiles =>
+      // For each bucket id, firstly we get all files belong to this bucket, by detecting bucket
+      // id from file name. Then read these files into a RDD(use one-partition empty RDD for empty
+      // bucket), and coalesce it to one partition. Finally union all bucket RDDs to one result.
+      val perBucketRows = (0 until maybeBucketSpec.get.numBuckets).map { bucketId =>
+        groupedBucketFiles.get(bucketId).map { inputStatuses =>
+          buildInternalScan(requiredColumns, filters, inputStatuses, broadcastedConf).coalesce(1)
```

Comment:
using coalesce and then Union here doesn't seem like a great way to do it. can we just create a single RDD, and each partition of the RDD just reads all the files? We would likely need to create a variant of the hadooprdd, but I think it's worth doing it.